src/js/actions/layers.js

```diff
@@ -768,55 +769,87 @@ define(function (require, exports) {
             return Promise.resolve();
         }
 
+        if (modifier !== undefined && layerSpec.size > 1) {
+            return Promise.reject(new Error("Select with modifier requires a single layer"));
+        }
+
+        var selected = document.layers.selected,
+            pivot = document.layers.pivot,
+            nextSelected,
+            nextPivot;
+
+        switch (modifier) {
+        case "select":
+            nextSelected = layerSpec;
+            nextPivot = layerSpec.first();
+            break;
+        case "deselect":
+            nextSelected = selected.filterNot(function (layer) {
+                return layerSpec.first().id === layer.id;
+            });
+            if (pivot) {
+                if (pivot.id === layerSpec.first().id) {
+                    nextPivot = selected.first();
+                } else {
+                    nextPivot = pivot;
+                }
+            }
+            break;
+        case "add":
+            nextSelected = selected.concat(layerSpec);
+            nextPivot = pivot;
+            break;
+        case "addUpTo":
+            if (selected.isEmpty()) {
+                nextSelected = layerSpec;
+                nextPivot = layerSpec.first();
+            } else {
+                nextPivot = pivot || selected.first();
+
+                var pivotIndex = document.layers.indexOf(nextPivot),
+                    targetIndex = document.layers.indexOf(layerSpec.first()),
+                    firstIndex = Math.min(pivotIndex, targetIndex),
+                    lastIndex = Math.max(pivotIndex, targetIndex);
+
+                nextSelected = document.layers.exposed.filter(function (layer) {
+                    var index = document.layers.indexOf(layer);
+
+                    return firstIndex <= index && index <= lastIndex;
+                });
+            }
+            break;
+        default:
+            nextSelected = layerSpec;
+            nextPivot = null;
```

Comment:
if "select" is the default modifier, then it seems weird they are treated slightly differently (nextPivot assignment).  Maybe there's a good reason, just checking.